### PR TITLE
ip: clean resources on iface deletion

### PR DESCRIPTION
**Pull request by christophefontaine**
On iface deletion, delete assoticated routes.
specify module destructor priorities to ensure
resources are not freed when accessing them.

### CODE REVIEW

File: modules/ip/control/nexthop.c [hunk 1]
```diff
@@ -74,6 +74,17 @@ int ip4_nexthop_add(uint16_t vrf_id, ip4_addr_t ip, uint32_t *idx, struct nextho
...
+
+		rte_spinlock_lock(&nh->lock);
+		// Flush all held packets.
+		struct rte_mbuf *m = nh->held_pkts_head;
+		while (m != NULL) {
+			struct rte_mbuf *next = queue_mbuf_data(m)->next;
+			rte_pktmbuf_free(m);
+			m = next;
+		}
+		rte_spinlock_unlock(&nh->lock);
```
Comment (rjarry) [inline modules/ip/control/nexthop.c:86]: Can you split this into a separate patch? This seems more like a fix related to nexthops.

File: modules/ip/control/address.c [hunk 2]
```diff
@@ -142,6 +142,22 @@ static struct api_out addr_list(const void *request, void **response) {
 	return api_out(0, len);
 }
 
+void route4_cleanup(struct iface *iface);
+void nh4_cleanup(struct iface *iface);
```
Comment (rjarry) [inline modules/ip/control/address.c:146]: These two declarations seem not used.

File: modules/ip/control/route.c [hunk 3]
```diff
@@ -301,6 +301,76 @@ static void route4_fini(struct event_base *) {
...
+			    ((uint8_t *)&ip)[2],
+			    ((uint8_t *)&ip)[3],
+			    prefixlen,
+			    ((uint8_t *)&nh->ip)[0],
+			    ((uint8_t *)&nh->ip)[1],
+			    ((uint8_t *)&nh->ip)[2],
+			    ((uint8_t *)&nh->ip)[3]);
+
+			ip4_route_delete(vrf_id, ip, prefixlen);
+			ip4_route_delete(vrf_id, ip, 32);
```
Comment (rjarry) [inline modules/ip/control/route.c:340]: I don't know if it is safe to delete while iterating.

File: modules/ip/control/route.c [hunk 4]
```diff
@@ -301,6 +301,76 @@ static void route4_fini(struct event_base *) {
...
+			    __func__,
+			    ((uint8_t *)&ip)[0],
+			    ((uint8_t *)&ip)[1],
+			    ((uint8_t *)&ip)[2],
+			    ((uint8_t *)&ip)[3],
+			    prefixlen,
+			    ((uint8_t *)&nh->ip)[0],
+			    ((uint8_t *)&nh->ip)[1],
+			    ((uint8_t *)&nh->ip)[2],
+			    ((uint8_t *)&nh->ip)[3]);
```
Comment (rjarry) [inline modules/ip/control/route.c:337]: Logging IP and ethernet addresses seem quite frequent. We need wrappers around `rte_log` and `snprintf` to reduce boilerplate. E.g.:

```diff
-LOG(DEBUG,
-    "%s ip4_route_delete %d.%d.%d.%d/%d via %d.%d.%d.%d",
-    __func__,
-    ((uint8_t *)&ip)[0],
-    ((uint8_t *)&ip)[1],
-    ((uint8_t *)&ip)[2],
-    ((uint8_t *)&ip)[3],
-    prefixlen,
-    ((uint8_t *)&nh->ip)[0],
-    ((uint8_t *)&nh->ip)[1],
-    ((uint8_t *)&nh->ip)[2],
-    ((uint8_t *)&nh->ip)[3]);
+ LOG(DEBUG, "%4/%hhu via %4", &ip, &nh->ip);
```

Or:

```diff
-scols_line_sprintf(line, 2, ETH_ADDR_FMT, ETH_BYTES_SPLIT(nh->mac.bytes));
+scols_line_sprintf(line, 2, "%2", &nh->mac));
```

What do you think?

**Review by rjarry — CHANGES_REQUESTED** [review]

File: modules/ip/control/nexthop.c (hunk 1 above)
Comment (christophefontaine) [reply modules/ip/control/nexthop.c:86]: indeed.

File: modules/ip/control/route.c (hunk 3 above)
Comment (christophefontaine) [reply modules/ip/control/route.c:334]: I guess this is safe as we iterate over ribs, but ip4_route_delete affects a fib.

**Review by rjarry — APPROVED** [review]
